### PR TITLE
Mypy: add static type checking for python to travis

### PR DESCRIPTION
**Pull request by trws**
This is dependent on #2805, but ended up essentially ready to go as soon as that goes forward.

### CODE REVIEW

**Comment by trws** [comment]
@Mergifyio rebase

**Comment by trws** [comment]
Well @grondo, there was definitely some funny business in barrier that could have caused issues.  Have a fix here, but have yet to figure out what the issue is with job info.

**Comment by trws** [comment]
Ok, we're green here.  Anyone willing to give me a MWP? @SteVwonder, @grondo?

**Comment by grondo** [comment]
I have a couple very minor testsuite fixes, we could put together PR if you want to split python test fixes out of this possibly unrelated PR.

File: .travis.yml [hunk 1]
```diff
@@ -43,7 +53,7 @@ jobs:
        - CXX=g++-8
        - ARGS="--with-flux-security --enable-sanitizer=address"
        - PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/8/libasan.so
-       - ASAN_OPTIONS=detect_leaks=0 # TODO: fix/suppress tests so we can turn this on
+       - ASAN_OPTIONS=detect_leaks=0,start_deactivated=true # TODO: fix/suppress tests so we can turn this on
```
Comment (SteVwonder) [inline .travis.yml]: Out of curiosity, what was the motivation for this flag?  Based on the docs, it looks like it delays allocating memory, but doesn't necessarily reduce memory footprint overall.  [In the docs](https://github.com/google/sanitizers/wiki/AddressSanitizerFlags), the line about this that stood out to me was "This is mainly intended to be used on Android.".

File: .travis.yml [hunk 2]
```diff
@@ -25,6 +25,16 @@ jobs:
       deploy: skip
       after_success: skip
       after_failure: skip
+    - stage: 'style checks'
+      name: 'mypy'
+      language: 'python'
+      python: '3.6'
+      install: python3 -m pip install 'mypy' --force-reinstall
+      script: mypy --config-file=src/bindings/python/.mypy.ini
```
Comment (SteVwonder) [inline .travis.yml]: Does this run on the entire codebase (i.e., does it include `src/cmd`), or just on the bindings?  If the former, can we have it run on the front-end commands too?  If the latter, do we want to create a script like `./scripts/pylint` for mypy, to make running locally easier? (not that this is hard to remember, but maybe having all the linters and formatters in the same place will make it easier to remember to run).

File: .travis.yml (hunk 1 above)
Comment (grondo) [reply .travis.yml]: @trws nicely described the reason in his commit message:
```
This causes asan to start in a low memory/low overhead mode until
instrumented code is loaded into the process.  In my environment this
lowers the asan run time for our full test suite by about 10%.
```

Comment (grondo) [reply .travis.yml]: Though looking at historical runtime, not sure it had an effect in the travis environment. :man_shrugging:

Comment (SteVwonder) [reply .travis.yml]: Thanks @grondo!  Should've gone through the commits one-by-one. :man_facepalming:

File: src/bindings/python/.mypy.ini [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[mypy]
+python_version = 3.6
+files=src/bindings/python/flux
```
Comment (SteVwonder) [inline src/bindings/python/.mypy.ini]: Do we want mypy to run on the front-end commands in `src/cmd` too?

File: .travis.yml (hunk 2 above)
Comment (trws) [reply .travis.yml]: Let me check something, if I move the mypy.ini to the root it should just be "run mypy with no args" which would make it easier to cover the whole project, I'll do one of the two in a commit in a bit.

File: src/bindings/python/.mypy.ini (hunk 3 above)
Comment (trws) [reply src/bindings/python/.mypy.ini]: Yeah, we probably do, and the tests also.  We do not want it to run on _flux though, I'll tweak that as well, thanks for noticing this.

File: .travis.yml (hunk 2 above)
Comment (trws) [reply .travis.yml]: mypy.ini is now at the root, just run `mypy` to check essentially everything.

File: src/bindings/python/.mypy.ini (hunk 3 above)
Comment (trws) [reply src/bindings/python/.mypy.ini]: We now check almost everything, need a way to generate stubs for constants to get the last two files to type check.  I can generate these easily after a build of the python module by using the module itself, but getting it in place without building the full tree is much less straightforward.  Mulling over how to deal with this.

File: t/t2204-job-info.t [hunk 4]
```diff
@@ -16,29 +16,33 @@ if test "$TEST_LONG" = "t"; then
     test_set_prereq LONGTEST
 fi
 
+fj_wait_event() {
```
Comment (SteVwonder) [inline t/t2204-job-info.t]: Assuming this change doesn't become it's own PR (per the slack discussion), I'd recommend making it it's own commit in this PR.

Comment (trws) [reply t/t2204-job-info.t]: I did not mean for this change to land here... and there is more fixup to actually make it work, alternate PR coming.

Comment (trws) [reply t/t2204-job-info.t]: I think I got this removed from the PR, and moved various other test fixes over to #2849.  Hopefully that still lets this one pass, we shall see.

**Pull request merged**